src/raftstore/store/snap.rs

```diff
@@ -109,35 +114,46 @@ impl Display for SnapKey {
     }
 }
 
-pub struct ApplyContext {
+#[derive(Default)]
+pub struct BuildContext {
+    pub snapshot_size: u64,
+    pub snapshot_kv_count: usize,
+}
+
+impl BuildContext {
+    pub fn new() -> BuildContext {
+        BuildContext { ..Default::default() }
+    }
+}
+
+pub struct ApplyOptions {
     pub db: Arc<DB>,
     pub region: Region,
     pub abort: Arc<AtomicUsize>,
     pub write_batch_size: usize,
-    pub snapshot_size: usize,
-    pub snapshot_kv_count: usize,
 }
 
 /// `Snapshot` is a trait for snapshot.
 /// It's used in these scenarios:
 ///   1. build local snapshot
-///   1. read local snapshot and then replicate it to remote raftstores
-///   2. receive snapshot from remote raftstore and write it to local storage
-///   3. snapshot gc
+///   2. read local snapshot and then replicate it to remote raftstores
+///   3. receive snapshot from remote raftstore and write it to local storage
 ///   4. apply snapshot
+///   5. snapshot gc
 pub trait Snapshot: Read + Write + Send {
     fn build(&mut self,
              snap: &DbSnapshot,
              region: &Region,
-             snap_data: &mut RaftSnapshotData)
-             -> raft::Result<()>;
+             snap_data: &mut RaftSnapshotData,
+             context: &mut BuildContext)
+             -> RaftStoreResult<()>;
     fn path(&self) -> &str;
     fn exists(&self) -> bool;
     fn delete(&self);
     fn meta(&self) -> io::Result<Metadata>;
     fn total_size(&self) -> io::Result<u64>;
     fn save(&mut self) -> io::Result<()>;
-    fn apply(&mut self, context: &mut ApplyContext) -> Result<()>;
+    fn apply(&mut self, options: ApplyOptions) -> Result<()>;
```

Comment:
`Context` is for a bunch of arguments that are modified inside the calling function. `Options` are for these not to be modified.